File: src-tauri/src/engine/app_context.rs
//! Single source of truth for “what kind of app is this?”
//!
//! Both the feature engine and snapback tracker read from here so they never
//! disagree about whether Slack is work and YouTube is a distraction.

/// Flags describing the active window. `Copy` means small structs can be
/// passed by value cheaply (like copying a few bools on the stack).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct AppContext {
    pub is_browser: bool,
    pub is_ide: bool,
    pub is_communication: bool,
    pub is_entertainment: bool,
    pub is_productivity: bool,
    pub is_terminal: bool,
    /// True when the window *title* names a known distraction (e.g. “YouTube”).
    pub title_is_distracting: bool,
}

impl AppContext {
    pub fn unknown() -> Self {
        Self {
            is_browser: false,
            is_ide: false,
            is_communication: false,
            is_entertainment: false,
            is_productivity: false,
            is_terminal: false,
            title_is_distracting: false,
        }
    }

    pub fn productivity_category(&self) -> &'static str {
        if self.is_ide {
            "Building"
        } else if self.is_productivity {
            "Writing"
        } else if self.is_browser {
            "Browsing"
        } else if self.is_communication {
            "Communicating"
        } else if self.is_entertainment {
            "Entertainment"
        } else {
            "Unknown"
        }
    }
}

const BROWSERS: &[&str] = &["chrome", "msedge", "firefox", "brave", "opera", "safari"];
const IDES: &[&str] = &[
    "code", "cursor", "devenv", "idea", "pycharm", "clion", "rider", "xcode",
];
const COMMUNICATION: &[&str] = &["slack", "discord", "teams", "outlook", "zoom", "messages"];
const ENTERTAINMENT_APPS: &[&str] = &["spotify", "steam", "vlc", "netflix", "youtube"];
const PRODUCTIVITY: &[&str] = &[
    "word", "excel", "powerpnt", "notion", "obsidian", "pages", "figma",
];
const TERMINALS: &[&str] = &["terminal", "iterm", "warp", "alacritty", "kitty", "wezterm"];

/// Title keywords catch distractions inside browsers (e.g. “YouTube — Chrome”).
const DISTRACTING_TITLE_KEYWORDS: &[&str] = &[
    "youtube", "netflix", "twitter", "reddit", "instagram", "tiktok", "twitch",
    "facebook", "hulu", "disney+", "prime video",
];

/// Classify from app name + window title. Both strings are borrowed (`&str`) —
/// we only read them, we don't take ownership (no heap allocation for the names).
pub fn classify(app_name: &str, window_title: &str) -> AppContext {
    let name = app_name.to_lowercase();
    let title = window_title.to_lowercase();

    let title_is_distracting = DISTRACTING_TITLE_KEYWORDS
        .iter()
        .any(|kw| title.contains(kw));

    AppContext {
        is_browser: BROWSERS.iter().any(|b| name.contains(b)),
        is_ide: IDES.iter().any(|b| name.contains(b)),
        is_communication: COMMUNICATION.iter().any(|b| name.contains(b)),
        is_entertainment: ENTERTAINMENT_APPS.iter().any(|b| name.contains(b)),
        is_productivity: PRODUCTIVITY.iter().any(|b| name.contains(b)),
        is_terminal: TERMINALS.iter().any(|b| name.contains(b)),
        title_is_distracting,
    }
}

/// Clearly not work — entertainment app or distracting site in the title.
pub fn is_clearly_off_task(ctx: &AppContext) -> bool {
    ctx.is_entertainment || ctx.title_is_distracting
}

/// Should snapback treat this moment as “on task” (worth saving / returning to)?
///
/// `focus_state` comes from the classifier (`Option` = maybe not computed yet).
/// `Option<&str>` is a maybe-present string slice — like `Optional[str]` in Python.
pub fn snapback_on_task(
    ctx: &AppContext,
    window_title: &str,
    focus_state: Option<&str>,
    session_goal: Option<&str>,
) -> bool {
    if is_clearly_off_task(ctx) {
        return false;
    }
    if focus_state == Some("DISTRACTED") {
        return false;
    }

    if let Some(goal) = session_goal.filter(|g| !g.trim().is_empty()) {
        let alignment = crate::engine::goal_alignment::alignment_score(goal, ctx, window_title);
        if alignment >= 0.72 {
            return true;
        }
        if alignment <= 0.35 {
            return false;
        }
    }

    if ctx.is_ide || ctx.is_productivity || ctx.is_terminal {
        return true;
    }

    if ctx.is_communication {
        return true;
    }

    match focus_state {
        Some("PRODUCTIVE") | Some("DEEP_FOCUS") | Some("PSEUDO_PRODUCTIVE") => true,
        _ => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn cursor_is_ide_and_on_task() {
        let ctx = classify("Cursor", "classifier.rs — Snapback");
        assert!(ctx.is_ide);
        assert!(snapback_on_task(&ctx, "classifier.rs — Snapback", None, None));
    }

    #[test]
    fn youtube_in_browser_title_is_off_task() {
        let ctx = classify("Google Chrome", "Rick Astley - YouTube");
        assert!(ctx.is_browser);
        assert!(ctx.title_is_distracting);
        assert!(!snapback_on_task(&ctx, "Rick Astley - YouTube", Some("PRODUCTIVE"), None));
    }

    #[test]
    fn classifier_distracted_overrides_slack() {
        let ctx = classify("Slack", "#general");
        assert!(ctx.is_communication);
        assert!(snapback_on_task(&ctx, "#general", Some("PRODUCTIVE"), None));
        assert!(!snapback_on_task(&ctx, "#general", Some("DISTRACTED"), None));
    }

    #[test]
    fn generic_browser_needs_classifier_or_stays_off_task() {
        let ctx = classify("Google Chrome", "New Tab");
        assert!(ctx.is_browser);
        assert!(!snapback_on_task(&ctx, "New Tab", None, None));
        assert!(snapback_on_task(&ctx, "New Tab", Some("PRODUCTIVE"), None));
    }

    #[test]
    fn research_goal_makes_docs_browser_on_task() {
        let ctx = classify("Google Chrome", "Rust documentation - std");
        assert!(snapback_on_task(
            &ctx,
            "Rust documentation",
            None,
            Some("research tokio docs"),
        ));
    }
}

File: src/engine/app_context.cpp
#include "engine/app_context.hpp"

#include <algorithm>
#include <array>
#include <cctype>

namespace snapback {
namespace {

constexpr std::array<const char*, 6> kBrowsers = {
    "chrome", "msedge", "firefox", "brave", "opera", "safari"};
constexpr std::array<const char*, 8> kIdes = {
    "code", "cursor", "devenv", "idea", "pycharm", "clion", "rider", "xcode"};
constexpr std::array<const char*, 6> kCommunication = {
    "slack", "discord", "teams", "outlook", "zoom", "messages"};
constexpr std::array<const char*, 5> kEntertainmentApps = {
    "spotify", "steam", "vlc", "netflix", "youtube"};
constexpr std::array<const char*, 7> kProductivity = {
    "word", "excel", "powerpnt", "notion", "obsidian", "pages", "figma"};
constexpr std::array<const char*, 6> kTerminals = {
    "terminal", "iterm", "warp", "alacritty", "kitty", "wezterm"};
constexpr std::array<const char*, 11> kDistractingTitleKeywords = {
    "youtube", "netflix", "twitter", "reddit", "instagram", "tiktok",
    "twitch", "facebook", "hulu", "disney+", "prime video"};

std::string lower(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(),
                   [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
    return s;
}

template <std::size_t N>
bool contains_any(const std::string& value, const std::array<const char*, N>& needles) {
    return std::any_of(needles.begin(), needles.end(),
                       [&](const char* needle) { return value.find(needle) != std::string::npos; });
}

bool matches_rule_pattern(const std::string& pattern,
                          const std::string& app_name,
                          const std::string& window_title) {
    return app_name.find(pattern) != std::string::npos ||
           window_title.find(pattern) != std::string::npos;
}

double category_alignment(const std::string& category,
                          const AppContext& ctx,
                          const std::string& title) {
    const std::string name = lower(category);
    const std::string t = lower(title);
    if (ctx.is_ide || ctx.is_terminal) return 0.95;
    if (ctx.is_productivity) return 0.95;
    if (ctx.is_browser && !ctx.title_is_distracting) {
        return name.find("research") != std::string::npos || name.find("read") != std::string::npos
                   ? 0.85
                   : (t.find("github") != std::string::npos || t.find("docs") != std::string::npos
                          ? 0.8
                          : 0.55);
    }
    if (ctx.is_communication) {
        return name.find("commun") != std::string::npos || name.find("meeting") != std::string::npos
                   ? 0.9
                   : 0.35;
    }
    return 0.5;
}

}  // namespace

std::vector<GoalCategory> default_goal_categories() {
    return {
        {"coding", {"code", "coding", "bug", "fix", "implement", "rust", "api", "refactor",
                     "test", "debug", "build", "feature", "compile", "merge", "pr", "pull request"}},
        {"writing", {"write", "writing", "report", "essay", "document", "blog", "draft", "paper", "notes"}},
        {"research", {"research", "read", "reading", "learn", "study", "article", "docs", "documentation"}},
        {"communication", {"email", "meeting", "call", "slack", "message", "reply", "interview", "present"}},
    };
}

double goal_alignment_score(const std::optional<std::string>& goal,
                            const AppContext& ctx,
                            const std::string& title,
                            const std::vector<GoalCategory>& categories) {
    const std::string g = lower(goal.value_or(""));
    if (g.empty()) return 0.5;
    const auto active = categories.empty() ? default_goal_categories() : categories;
    double best = 0.5;
    for (const auto& category : active) {
        const auto category_name = lower(category.name);
        const bool matches = std::any_of(category.keywords.begin(), category.keywords.end(),
                                         [&](const auto& keyword) {
                                             return !keyword.empty() && g.find(lower(keyword)) != std::string::npos;
                                         });
        if (!matches) continue;
        double score = category_alignment(category_name, ctx, title);
        if (category_name.find("communication") != std::string::npos && ctx.is_browser &&
            lower(title).find("mail") != std::string::npos) score = 0.85;
        best = std::max(best, score);
    }
    return std::clamp(best, 0.0, 1.0);
}

std::string AppContext::productivity_category() const {
    if (personal_block) return "Entertainment";
    if (is_ide) return "Building";
    if (is_productivity || personal_allow) return "Writing";
    if (is_browser) return "Browsing";
    if (is_communication) return "Communicating";
    if (is_entertainment) return "Entertainment";
    return "Unknown";
}

AppContext classify_app_context(const std::string& app_name,
                                const std::string& window_title,
                                const std::vector<AppRuleRecord>& rules) {
    const std::string name = lower(app_name);
    const std::string title = lower(window_title);

    AppContext ctx;
    ctx.is_browser = contains_any(name, kBrowsers);
    ctx.is_ide = contains_any(name, kIdes);
    ctx.is_communication = contains_any(name, kCommunication);
    ctx.is_entertainment = contains_any(name, kEntertainmentApps);
    ctx.is_productivity = contains_any(name, kProductivity);
    ctx.is_terminal = contains_any(name, kTerminals);
    ctx.title_is_distracting = contains_any(title, kDistractingTitleKeywords);

    bool allow = false;
    bool block = false;
    for (const auto& rule : rules) {
        const std::string pattern = lower(rule.pattern);
        if (!matches_rule_pattern(pattern, name, title)) continue;
        if (rule.rule_type == AppRuleKind::Block) block = true;
        if (rule.rule_type == AppRuleKind::Allow) allow = true;
    }

    if (block) {
        ctx.personal_block = true;
        ctx.personal_allow = false;
        ctx.is_entertainment = true;
        ctx.title_is_distracting = true;
    } else if (allow) {
        ctx.personal_allow = true;
        ctx.is_entertainment = false;
        ctx.title_is_distracting = false;
        ctx.is_productivity = true;
    }
    return ctx;
}

bool is_clearly_off_task(const AppContext& ctx) {
    if (ctx.personal_allow) return false;
    return ctx.personal_block || ctx.is_entertainment || ctx.title_is_distracting;
}

bool snapback_on_task(const AppContext& ctx,
                      const std::string& window_title,
                      const std::optional<std::string>& focus_state,
                      const std::optional<std::string>& session_goal,
                      const std::vector<GoalCategory>& categories) {
    if (is_clearly_off_task(ctx)) return false;
    if (focus_state == std::optional<std::string>("DISTRACTED")) return false;
    if (ctx.personal_allow) return true;

    if (session_goal && !session_goal->empty()) {
        const double score = goal_alignment_score(session_goal, ctx, window_title, categories);
        if (score >= 0.72) return true;
        if (score <= 0.35) return false;
    }

    if (ctx.is_ide || ctx.is_productivity || ctx.is_terminal || ctx.is_communication) return true;
    if (focus_state == std::optional<std::string>("PRODUCTIVE") ||
        focus_state == std::optional<std::string>("DEEP_FOCUS") ||
        focus_state == std::optional<std::string>("PSEUDO_PRODUCTIVE")) return true;
    return false;
}

}  // namespace snapback
